### PR TITLE
test(passes): before/after IR coverage + Submit-awareness fixes (#1615)

### LINKED ISSUE #1615: [Pass Bug] 8 passes skip the Submit IR kind (exact-kind Call dispatch) inside pl.manual_scope
### Pass Name

Umbrella tracking issue across **8 passes** (shared root cause). Per-pass checklist below: `normalize_return_order`, `inject_gm_pipe_buffer`, `inline_functions`, `flatten_call_expr`, `convert_tensor_to_tile_ops`, `optimize_orch_tensors`, `materialize_tensor_strides`, `fuse_create_assemble_to_slice`.

### Description

Each of these passes walks call-like IR nodes via **exact-kind** `As<Call>` / `VisitExpr_(CallPtr)` / the `GetCallFromStmt` helper (`include/pypto/ir/transforms/utils/transform_utils.h:84`, which is `As<Call>`-only). `Submit` is a **sibling `ObjectKind`** (emitted by `pl.submit` inside `pl.manual_scope`), **not** a `Call` subclass, so these passes silently skip it — the exact failure mode documented in `.claude/rules/pass-submit-awareness.md`.

All 8 run **before** `DeriveCallDirections` (pass 34, the point where `Submit` is lowered to `Call`), so they genuinely see live `Submit` nodes. `DeriveCallDirections` itself was already made submit-aware in #1552 and is the reference implementation.

Each item below has a committed `xfail(strict=False)` regression test (branch `test/pass-before-after-coverage`) that encodes correct behavior and **fails today**; it flips to passing once its pass is fixed.

- [x] **normalize_return_order** — `normalize_return_order_pass.cpp:241` (`TupleIndexPermutationMutator`): `Submit` result `TupleGetItem` indices not remapped after Out-param reorder → unpacked names bind the wrong output buffers. Repro: `test_normalize_return_order.py::TestNormalizeReturnOrderSubmit::test_submit_swapped_returns_remapped`
- [x] **inject_gm_pipe_buffer** — uses `GetCallFromStmt`: a Group launched via `pl.submit` on Ascend910B is invisible → no `tensor.create` placeholder, `__gm_pipe_buffer` not forwarded. Repro: `test_inject_gm_pipe_buffer.py::test_gm_pipe_injection_handles_submit_launched_group`
- [x] **inline_functions** — `CalledInlineCollector`/`InlineCallsMutator`: `pl.submit` of an Inline fn is not inlined; the fn is dropped but the dangling `Submit` survives and `InlineFunctionsEliminated` misses it. Repro: `test_inline_functions.py::TestInlineFunctionsSubmitCallSite::test_submit_of_inline_eliminates_reference`
- [x] **flatten_call_expr** — no `VisitExpr_(SubmitPtr)`: a nested `Call` arg of a `Submit` is not hoisted → three-address invariant violated. Repro: `test_flatten_call_expr_pass.py::TestFlattenPreservesAttrs::test_nested_call_arg_in_submit_is_flattened`
- [x] **convert_tensor_to_tile_ops** — `CallSiteUpdateMutator`: a `pl.submit` call site to a transformed InCore gets no `tensor.create` and the appended Out arg is not forwarded. Repro: `test_convert_tensor_to_tile_ops.py::TestSubmitCallSiteUpdate::test_submit_call_site_gets_tensor_create`
- [x] **optimize_orch_tensors** — `OrchRewriter::TryRewriteCall`: the `IsSubmitCall` branch is dead for real `Submit` nodes → a windowable kernel launched via `pl.submit` stays baseline. Repro: `test_optimize_orch_tensors.py::TestOutWindowSubmitCall::test_submit_windowable_kernel_is_externalized`
- [x] **materialize_tensor_strides** — no `VisitExpr_(SubmitPtr)`: a `Submit` node's own return type keeps an empty stride while every other reachable `TensorType` is materialized. Repro: `test_materialize_tensor_strides_pass.py::test_submit_return_type_materialized`
- [x] **fuse_create_assemble_to_slice** — `BufferRootCollector` (`As<Call>`): a `pl.submit` callee never propagates its Out-param buffer root → create+assemble around a submit is left unfused. Repro: `test_fuse_create_assemble_to_slice.py::test_submit_create_assemble_fused_to_slice`

Related (context, different root cause): #1368.

### Git Commit ID

35731d4833cdef3c49d8da254bba1d9a145be92f

### Before IR (Input)

Representative case — `normalize_return_order`. The InCore `kernel` returns `(out_b, out_a)` (reversed vs Out-param order); the orchestrator launches it via `pl.submit` and unpacks `(a, b)`:

```python
@pl.program
class Before:
    @pl.function(type=pl.FunctionType.InCore)
    def kernel(self, x: pl.Tensor[[16], pl.FP32],
               out_a: pl.Out[pl.Tensor[[16], pl.FP32]],
               out_b: pl.Out[pl.Tensor[[16], pl.FP32]]) -> tuple[pl.Tensor[[16], pl.FP32], pl.Tensor[[16], pl.FP32]]:
        x_tile = pl.load(x, [0], [16])
        a_tile = pl.tile.add(x_tile, x_tile)
        b_tile = pl.tile.mul(x_tile, x_tile)
        out_b_store = pl.store(b_tile, [0], out_b)
        out_a_store = pl.store(a_tile, [0], out_a)
        return (out_b_store, out_a_store)   # reversed vs (out_a, out_b)

    @pl.function(type=pl.FunctionType.Orchestration)
    def main(self, x, out_a: pl.Out[...], out_b: pl.Out[...]) -> tuple[...]:
        with pl.manual_scope():
            (a, b), tid = pl.submit(self.kernel, x, out_a, out_b)
        return (a, b)
```

### Expected IR (After Transformation)

Step A reorders the kernel's returns to match Out-param order `(out_a, out_b)`; Step B must then **remap the caller's projections** so `a`/`b` still bind the same logical buffers (encoded as the swapped `(b, a)` unpack):

```python
class Expected:
    def kernel(...) -> tuple[...]:
        ...
        return (out_a_store, out_b_store)   # normalized order

    def main(...) -> tuple[...]:
        with pl.manual_scope():
            (b, a), tid = pl.submit(self.kernel, x, out_a, out_b)  # _submit_tmp indices remapped
        return (a, b)
```

### Actual IR or Error

Step A reorders the kernel returns, but `TupleIndexPermutationMutator` matches `As<Call>` only and never tracks the `Submit` result, so the caller's `TupleGetItem` indices are left unchanged:

```python
    def main(...) -> tuple[...]:
        with pl.manual_scope():
            (a, b), tid = pl.submit(self.kernel, x, out_a, out_b)  # NOT remapped — stale indices
        return (a, b)
```

`ir.assert_structural_equal(After, Expected)` fails on `['main'].body` (the manual_scope projection order differs). Net effect: after the kernel's outputs are reordered, `a` and `b` silently bind the wrong output tensors.

### NPU Kind (if relevant)

N/A (not hardware-specific)

### Host Platform (if relevant)

N/A (not hardware-specific)

### Additional Context

**Common fix shape** (same for all 8): make each call-walk Submit-aware — use `AsCallLike` / a `SubmitToCallView` adapter, preserve Submit-ness through rewrites, relax the args-vs-params size guard to the **prefix invariant** (`args_.size() <= params_.size()`, not `==`), and account for the trailing `Scalar[TASK_ID]` in the return tuple. See `.claude/rules/pass-submit-awareness.md` and the reference fix in #1552 (`DeriveCallDirections`).

The 8 `xfail(strict=False)` regression tests are committed on branch `test/pass-before-after-coverage`; each will flip to passing (xpass) once its pass is fixed, so they double as ready-made acceptance checks.

---

**Status: implemented** on branch `test/pass-before-after-coverage` (commit `b24d7409`). All 8 passes are now Submit-aware; the 8 committed before/after regression tests (previously `xfail`) pass as normal tests, and the full `tests/ut/ir` transforms suite is green. `DeriveCallDirections` reference fix: #1552.

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

Two related changes, in order:

1. **`test(passes)`** — add the project's canonical Before/After/Expected
   (`assert_structural_equal`) coverage across the HIGH+MEDIUM-priority transform
   passes: ~109 new structural tests + 16 conversions of weak property/smoke
   tests, targeting previously-uncovered branches (`pl.manual_scope`/`pl.submit`
   handling, control-flow bodies, backend gates, edge cases).

2. **`fix(passes)` — Fixes #1615** — the new tests surfaced a family of 8
   "submit-awareness" bugs: `pl.submit` (inside `pl.manual_scope`) emits the
   `Submit` IR kind, a *sibling* of `Call` (not a subclass), and 8 passes walked
   call-like nodes via exact-kind `As<Call>` / `GetCallFromStmt` and silently
   skipped `Submit`. Each is now Submit-aware, preserving Submit-ness
   (`deps_`/`kwargs_`/`attrs_` and the trailing `Scalar[TASK_ID]` return) where
   it rewrites the node.

## The 8 fixes (#1615)

| Pass | Fix |
| ---- | --- |
| `normalize_return_order` | track Submit results → remap `TupleGetItem` projection indices after Out-param reorder |
| `materialize_tensor_strides` | `VisitExpr_(SubmitPtr)` → materialize the Submit's own return-type strides |
| `flatten_call_expr` | `VisitExpr_(SubmitPtr)` → hoist nested Call/Submit args (preserve `deps_`) |
| `fuse_create_assemble_to_slice` | route Submit through the Out/InOut buffer-root analysis via `SubmitToCallView` |
| `inject_gm_pipe_buffer` | Submit-aware call graph + both rewriters + recurse `RuntimeScopeStmt` so a `pl.submit`-launched Group gets its `__gm_pipe_buffer` |
| `convert_tensor_to_tile_ops` | `HandleSubmitCallSite` → allocate + forward appended runtime-Out args at a Submit call site |
| `optimize_orch_tensors` | dispatch `TryRewriteCall` on the Submit's Call view, rebuild a Submit for window-write externalization |
| `InlineFunctionsEliminated` verifier | flag a dangling `pl.submit` to a dropped Inline function |

Reference implementation: #1552 (`DeriveCallDirections`). The 8 committed
before/after regression tests (initially `xfail`) now pass as normal tests.

## Verification

- `tests/ut/ir/` + `tests/ut/pass/`: **3680 passed, 3 skipped, 0 failed**.
- clang-tidy, clang-format, cpplint, ruff, pyright all clean.
- Test-only commit changes no source; the fix commit changes no IR definitions
  or bindings (pass-internal only).

### CODE REVIEW

File: src/ir/transforms/inject_gm_pipe_buffer_pass.cpp [hunk 1]
```diff
@@ -145,26 +159,41 @@ StmtPtr RewriteCallsForGMBuffer(const StmtPtr& body, const std::unordered_set<st
...
-      return new_call;
+      if (auto call = As<Call>(value)) {
+        auto new_call = MutableCopy(call);
+        new_call->args_.push_back(gm_param);
+        return new_call;
+      }
+      auto new_submit = MutableCopy(As<Submit>(value));
+      new_submit->args_.push_back(gm_param);
+      return new_submit;
     };
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/inject_gm_pipe_buffer_pass.cpp:188]: Done in `ee7368ce` — cast `Call`/`Submit` once at the top of the lambda and reuse, dropping the redundant `As<>` calls.

File: src/ir/transforms/inject_gm_pipe_buffer_pass.cpp [hunk 2]
```diff
@@ -239,27 +280,42 @@ StmtPtr RewriteCallsWithPerCallGMBuffer(const StmtPtr& body,
...
+
+    if (auto call = As<Call>(value)) {
+      auto new_call = MutableCopy(call);
+      new_call->args_.push_back(gm_var);
+      return std::make_pair(create_stmt_ptr, ExprPtr(new_call));
+    }
+    auto new_submit = MutableCopy(As<Submit>(value));
+    new_submit->args_.push_back(gm_var);
+    return std::make_pair(create_stmt_ptr, ExprPtr(new_submit));
   };
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/inject_gm_pipe_buffer_pass.cpp:318]: Done in `ee7368ce` — same cast-once cleanup applied to this lambda.

File: src/ir/transforms/optimize_orch_tensors_pass.cpp [hunk 3]
```diff
@@ -2472,8 +2479,22 @@ class OutWindowExternalizer {
...
+        submit_attrs.reserve(new_attrs.size());
+        for (const auto& [k, v] : new_attrs) {
+          if (k != kAttrManualDepEdges) submit_attrs.emplace_back(k, v);
+        }
+        new_call = std::make_shared<Submit>(cloned_gvar, new_args, submit->deps_, call->kwargs_,
+                                            std::move(submit_attrs), new_return_type, submit->span_);
+      } else {
+        new_call = std::make_shared<Call>(cloned_gvar, new_args, call->kwargs_, new_attrs, new_return_type,
+                                          call->span_);
+      }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/optimize_orch_tensors_pass.cpp:2512]: Done in `ee7368ce` — the windowed `Submit` is now rebuilt with `submit->kwargs_` directly rather than the `SubmitToCallView`'s `call->kwargs_`.

File: src/ir/transforms/materialize_tensor_strides_pass.cpp [hunk 4]
```diff
@@ -216,11 +216,37 @@ class MaterializeTensorStridesMutator : public IRMutator {
...
+    auto base = IRMutator::VisitExpr_(op);
+    auto submit = As<Submit>(base);
+    if (!submit) return base;
+    auto new_return_type = MaterializeType(submit->GetType());
+    if (new_return_type.get() == submit->GetType().get()) return submit;
+    // MaterializeType recurses the TupleType, materializing each leading
+    // return TensorType and leaving the trailing Scalar[TASK_ID] untouched.
+    // Note the 7-arg Submit ctor order is (op, args, deps, kwargs, attrs, ...).
+    return std::make_shared<Submit>(submit->op_, submit->args_, submit->deps_, submit->kwargs_,
+                                    submit->attrs_, std::move(new_return_type), submit->span_);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/materialize_tensor_strides_pass.cpp:247]: Verified this is a non-issue: the `AssignStmt` LHS var's type is materialized via the pass's general type-visit, not only the Call-specific sync block. Post-pass the LHS var and the RHS `Submit` types are structurally equal (both DN stride `[1, 4]`), so there is no LHS/RHS mismatch.

File: tests/ut/ir/transforms/test_collect_comm_groups.py [hunk 5]
```diff
@@ -329,6 +378,32 @@ def host_orch(self):
...
+    """
+
+    @pl.program
+    class P:
+        @pl.function(level=pl.Level.HOST, role=pl.Role.Orchestrator)
+        def host_orch(self):
+            buf = pld.alloc_window_buffer(1024)  # noqa: F841  # never windowed
+            return 0
+
+    with pytest.raises(Exception, match="no pld.tensor.window materialisation"):
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/transforms/test_collect_comm_groups.py]: Fixed in `ee7368ce` — escaped the literal dots: `match=r"no pld\.tensor\.window materialisation"`.

File: tests/ut/ir/transforms/test_inject_gm_pipe_buffer.py [hunk 6]
```diff
@@ -36,6 +36,175 @@ def _run_inject(program: ir.Program) -> ir.Program:
...
+            return updated
+
+        @pl.function(type=pl.FunctionType.Orchestration)
+        def main(
+            self,
+            a: pl.Tensor[[16, 16], pl.FP16],
+            out: pl.Out[pl.Tensor[[16, 16], pl.FP16]],
+        ) -> pl.Tensor[[16, 16], pl.FP16]:
+            with pl.manual_scope():
+                updated, tid = pl.submit(self.group_func, a, out)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/transforms/test_inject_gm_pipe_buffer.py]: Fixed in `ee7368ce` — prefixed the unused submit task-id with `_` (`_tid`).

File: tests/ut/ir/transforms/test_materialize_tensor_strides_pass.py [hunk 7]
```diff
@@ -303,5 +319,114 @@ def f(
...
+    # A function whose single return is a Tuple of two empty-stride DN tensors.
+    # Both elements must be materialized to their packed DN canonical stride:
+    #   [4, 8]    -> [1, 4]
+    #   [2, 4, 8] -> [32, 1, 4]
+    # (DN formula, doc 27-materialize_tensor_strides.md "Stride Formulas".)
+    def build(stride_2d, stride_3d):
+        x = ir.Var("x", _dn_tensor([4, 8], stride_2d), _SPAN)
+        ret_tuple = ir.TupleType([_dn_tensor([4, 8], stride_2d), _dn_tensor([2, 4, 8], stride_3d)])
+        body = ir.ReturnStmt([x, x], _SPAN)
+        func = ir.Function("f", [x], [ret_tuple], body, _SPAN)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/transforms/test_materialize_tensor_strides_pass.py]: Fixed in `ee7368ce` — the second `ReturnStmt` operand now uses a `[2, 4, 8]` var matching the declared tuple-return type (was reusing the `[4, 8]` var).

**Comment by Hzfengsy** [comment]
@codex review

File: src/ir/transforms/optimize_orch_tensors_pass.cpp [hunk 8]
```diff
@@ -2392,7 +2392,14 @@ class OutWindowExternalizer {
...
     std::optional<RewriteBundle> TryRewriteCall(const AssignStmtPtr& call_assign) {
-      auto call = As<Call>(call_assign->value_);
+      // Submit (pl.submit inside pl.manual_scope) is a sibling call-like kind;
+      // run the windowing analysis/rewrite on its augmented-Call view, then
+      // rebuild as a Submit to preserve task-launch semantics + deps_
+      // (.claude/rules/pass-submit-awareness.md). The per-callee analysis and
+      // windowed clone are callee-body-driven (Analyze() over all functions),
+      // so they exist regardless of the call-site kind.
+      auto submit = As<Submit>(call_assign->value_);
+      auto call = submit ? SubmitToCallView(submit) : As<Call>(call_assign->value_);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/optimize_orch_tensors_pass.cpp:2417]: Fixed in `fe90dd60` — `AddFullRootReadsFromStmt` now routes assign/eval values through a Call/Submit-aware helper (`AddFullRootReadsFromCallLike` → `SubmitToCallView`) and recurses `RuntimeScopeStmt`, so a later `pl.submit` reader of a windowed output is counted by the safety guard. Added a regression test (`test_submit_windowable_suppressed_by_later_full_submit_read`): a windowed submit whose full output is read by a later submit stays baseline.

**Review by lyfne123 — APPROVED** [review]
